umapi_client/connection.py

```diff
@@ -475,3 +546,84 @@ def call():
         if self.logger: self.logger.error("UMAPI timeout...giving up after %d attempts (%d seconds).",
                                           self.retry_max_attempts, total_time)
         raise UnavailableError(self.retry_max_attempts, total_time, result)
+
+
+class SessionManager:
+    """
+    Handles the session state for the connector.  Allows for variable session life, session
+    regeneration, and the option to disable sessions (connection pooling)
+    """
+    def __init__(self,
+                 user_agent,
+                 connection_pooling,
+                 session_max_age):
+
+        """
+        :param user_agent: [string] the user agent to be set in the request headers
+        :param connection_pooling: [bool] determines whether sessions will be reused
+        :param session_max_age: [int] age in seconds before session is scrapped and replaced
+        """
+
+        self.session = None
+        self.logger = logging.getLogger("umapi - SessionManager")
+        self.connection_pooling = connection_pooling
+        self.session_max_age = session_max_age
+        self.headers = {
+            'User-Agent': user_agent}
+
+        # Set the kind of requests made with the request_handler (request or session)
+        if self.connection_pooling:
+            self.update_session()
+        else:
+            self.request_handler = requests
+
+    # get, post and delete are pass through methods which make the desired call using the
+    # request_handler.  Allows to abstract the kind of call and reduces the needed methods
+    # from 6 to 3.  Each call includes session validation as a first step.
+    def get(self, url, **kwargs):
+        self.validate_session()
+        return self.request_handler.get(url, **kwargs)
+
+    def post(self, url, **kwargs):
+        self.validate_session()
+        return self.request_handler.post(url, **kwargs)
+
+    def delete(self, url, **kwargs):
+        self.validate_session()
+        return self.request_handler.delete(url, **kwargs)
+
+    def update_session(self):
+        """
+        Closes and removes the existing session from memory
+        Creates a fresh session object and adds the UA headers
+        """
+
+        if self.session:
+            self.session.close()
+            del self.session
+
+        session = requests.Session()
+        session.headers.update(self.headers)
+
+        # Session ID is the memory address of the session
+        self.session_id = id(session)
```

Comment:
You are not using this except in the log, so why are you keeping it in a member variable?  Just call id(session) when you write the log.  Memory locations are really bad things to use in logs, by the way, why not just do repr(session) or str(session) and let the requests library do a better job of reporting the session characteristics?

Reply:
The actual reasoning behind this was explicitly to check the memory location, to verify that it was indeed a new object.   Calling requests.Session() seems to hand over the same object if called in a row.  Without knowing the details of what that object actually contains, my intent was to create a definitely new object upon demand.  This was the same reason for using the "del" command in the other comment.  But I am glad to dig a little deeper on this and see if I can identify what's going on behind the scenes.